### PR TITLE
Ruff linting etc

### PR DESCRIPTION
**Pull request by hyanwong**
Hi @davidebbo - this is a mega PR that puts all the files into the format that `ruff` likes. It has picked up quite a few oddities and errors, which is good. I'm afraid I might have broken some stuff though, and since test coverage isn't complete, I might not pick it up.

Could you possible try out this PR with e.g. some simple tree-building code and check it's still producing the same results?

### CODE REVIEW

**Comment by hyanwong** [comment]
You can install the ruff pre-commit by simply typing `pre-commit install` int he github repo, I think.

File: data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Batoids_Aschliman2012.PHY [hunk 1]
```diff
@@ -1,4 +1,5 @@
-[Batoids & dates from:Body plan convergence in the evolution of skates and rays (Chondrichthyes: Batoidea) 
+[Batoids & dates from:
+Body plan convergence in the evolution of skates and rays (Chondrichthyes: Batoidea)
```
Comment (davidebbo) [inline data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Batoids_Aschliman2012.PHY:2]: Why would ruff mess with a non-Python file like a .PHY file?

File: oz_tree_build/images_and_vernaculars/get_wiki_images.py [hunk 2]
```diff
@@ -54,8 +51,8 @@
 # Copied from OZTree/OZprivate/ServerScripts/Utilities/getEOL_crops.py
 def subdir_name(doID):
     """
-    Make a valid subdirectory name in which to save images, based on the last 3 chars of
-    the data object ID
+    Make a valid subdirectory name in which to save images, based on the last
```
Comment (davidebbo) [inline oz_tree_build/images_and_vernaculars/get_wiki_images.py:54]: That feels like an over-agressive line length. Maybe best to configure it to match what Black did, to reduce the churn?

File: oz_tree_build/images_and_vernaculars/get_wiki_images.py [hunk 3]
```diff
@@ -190,10 +188,10 @@ def get_vernaculars_by_language_from_json_item(json_item):
 
     # For each language:
     # - We keep all the vernaculars
-    # - If none are marked as preferred, the first non-referred will be marked as preferred
+    # - If none are marked as preferred, use the first non-preferred as preferred
     # - If multiple are marked as preferred, the first one will be kept as preferred
     for vernaculars in vernaculars_by_language.values():
-        vernaculars.sort(reverse = True, key = lambda v: v["preferred"])
+        vernaculars.sort(reverse=True, key=lambda v: v["preferred"])
```
Comment (davidebbo) [inline oz_tree_build/images_and_vernaculars/get_wiki_images.py:194]: This might be another configurable thing, if we wanted to stay closer to Black?

**Comment by davidebbo** [comment]
The challenge here is that this PR has a mix of Black vs Ruff changes, and of unrelated logic changes. That makes it a little hard to review the logic changes.

Would it make more sense to keep everything as Black formatter for the time being while we fine tune the logic, and then later make a simple Ruff reformat phase, which should be entirely harmless?

File: data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Batoids_Aschliman2012.PHY (hunk 1 above)
Comment (hyanwong) [reply data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Batoids_Aschliman2012.PHY:2]: Yes, I don't know. I think it was accidentally rolled in from another change. It hasn't touched any of the other PHY files, and the change looks OK, so I was leaving it in there.

File: oz_tree_build/images_and_vernaculars/get_wiki_images.py (hunk 2 above)
Comment (hyanwong) [reply oz_tree_build/images_and_vernaculars/get_wiki_images.py:54]: What did we have in `black`? The recommended is 89, but I find that too short. I'm happy to make it longer. How about e.g. 99? Or even 120?

File: oz_tree_build/images_and_vernaculars/get_wiki_images.py (hunk 3 above)
Comment (hyanwong) [reply oz_tree_build/images_and_vernaculars/get_wiki_images.py:194]: I'm not bothered. I prefer the tight equals (no spaces) in parameter settings, and ISTR that black used to do this too?

**Comment by hyanwong** [comment]
Yes, I take your point. Ruff identified the logic issues, which I had to fix. 

I guess I could run Black on the all the ruff formatted files, and we should get roughly back to where we were, but with all the fixes?

**Comment by davidebbo** [comment]
If I pull this branch with no changes, `pytest` fails with:

    get_wiki_images.py:get_wiki_images.py:533 Multiple results for 'Homo sapiens'

I think that's because I already have the real entry in there, and the test is adding another by the same name. Maybe we should use made up taxa for the tests to avoid clashes?

**Comment by davidebbo** [comment]
> I guess I could run Black on the all the ruff formatted files, and we should get roughly back to where we were, but with all the fixes?

Yes, I think that would make sense. This way, we don't lose all the work you did you fix flagged logic errors.

**Comment by hyanwong** [comment]
> > I guess I could run Black on the all the ruff formatted files, and we should get roughly back to where we were, but with all the fixes?
> 
> Yes, I think that would make sense. This way, we don't lose all the work you did you fix flagged logic errors.

Actually, I just tried this, and Black likes all the changes that ruff made, so it doesn't turn anything back to how it used to be!

**Comment by hyanwong** [comment]
I guess as a first pass, I could make the first 4 commits here into a separate PR, as they don't really do much in the way for formatting.

File: data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Batoids_Aschliman2012.PHY (hunk 1 above)
Comment (davidebbo) [reply data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Batoids_Aschliman2012.PHY:2]: Oh I see, so you had some deliberate PHY changes that are desirable, and unrelated to the Ruff change? If so, that's fine.

**Comment by hyanwong** [comment]
> If I pull this branch with no changes, `pytest` fails with:
> 
> ```
> get_wiki_images.py:get_wiki_images.py:533 Multiple results for 'Homo sapiens'
> ```
> 
> I think that's because I already have the real entry in there, and the test is adding another by the same name. Maybe we should use made up taxa for the tests to avoid clashes?

Good point. I think we only need this for the test where we try a name rather than an OTT, though. So you could try changing the line that says `sp_name = "Homo sapiens"` to `sp_name = "Thisisnota speciesname"` and see if it passes?

File: data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Batoids_Aschliman2012.PHY (hunk 1 above)
Comment (hyanwong) [reply data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Batoids_Aschliman2012.PHY:2]: The changes are only in the comment section of those 2 PHY files. I think they must have been in my text editor buffer and were flushed when I did something else, and accidentally got into the PR.

File: oz_tree_build/images_and_vernaculars/get_wiki_images.py (hunk 3 above)
Comment (davidebbo) [reply oz_tree_build/images_and_vernaculars/get_wiki_images.py:194]: Ok, I'm fine with changing it, it was more about reducing the churn. If we temporarily move back to Black to reduce to diff (per other discussion), we can later use the Ruff defaults.

**Comment by davidebbo** [comment]
> So you could try changing the line that says `sp_name = "Homo sapiens"` to `sp_name = "Thisisnota speciesname"` and see if it passes?

Yes, all passes with that name.

File: oz_tree_build/images_and_vernaculars/get_wiki_images.py (hunk 2 above)
Comment (davidebbo) [reply oz_tree_build/images_and_vernaculars/get_wiki_images.py:54]: I see, it appears that black was simply not doing anything to comment lines regardless of the length. Going to 120 sgtm. This should in practice revert some of the Ruff changes, like in this particular case.

Comment (hyanwong) [reply oz_tree_build/images_and_vernaculars/get_wiki_images.py:54]: Good point. If you like, I can rebase the PR and make the ruff changes 120 lines instead? I'll need to force push it all though

**Comment by davidebbo** [comment]
> Actually, I just tried this, and Black likes all the changes that ruff made, so it doesn't turn anything back to how it used to be!

Interesting. I think this happens for a mix of two reasons:
1. Ruff is stricter, so it changes things that Black would accept either way. Hence Black is not always reverting the Ruff change.
2. Some of the formatting changes are to code that you had just added and had never gone through formatting. I suppose I would need to look at the full diff from before all your changes to the latest here, rather than just this PR, which doesn't include those earlier changes.

File: oz_tree_build/images_and_vernaculars/get_wiki_images.py (hunk 2 above)
Comment (davidebbo) [reply oz_tree_build/images_and_vernaculars/get_wiki_images.py:54]: Force pushing on a dev branch is fine. It's only bad on main. Or I suppose you could reconfigure the formatter to 120 and get it in as a new commit.

I don't want to make it more complicated for you than it has to be. If needed, I can just ignore comments when reviewing. It's the code logic that needs careful review.

**Comment by davidebbo** [comment]
Actually, Black does revert some things for me. e.g. if I save `test_extract_trees.py`, `test_one_taxon_with_exclusion_and_two_ancestors` gets reformatted over 3 lines instead of 1. Same for many other places.

I guess one thing I could do is resave everything with Black, and push to a dummy branch just so I can get a cleaner diff with everything from before my change. We'd then discard that branch.

I can probably do all this Sunday or early next week, once I'm back.

File: oz_tree_build/images_and_vernaculars/get_wiki_images.py (hunk 2 above)
Comment (davidebbo) [reply oz_tree_build/images_and_vernaculars/get_wiki_images.py:54]: Thanks!

**Comment by davidebbo** [comment]
Here is a URL that shows the full diff from before all your changes to the latest: https://github.com/OneZoom/tree-build/compare/69cdebe7adabb26be3d16ac3f59f1ee6ece7aa09..WIKI-IMG

So there are a lot of changes, but at least a bunch of them are caused by Black vs Ruff formatting, which I can hopefully reverse for diffing purpose.

**Comment by hyanwong** [comment]
> Actually, Black does revert some things for me. e.g. if I save `test_extract_trees.py`, `test_one_taxon_with_exclusion_and_two_ancestors` gets reformatted over 3 lines instead of 1. Same for many other places.
> 
> I guess one thing I could do is resave everything with Black, and push to a dummy branch just so I can get a cleaner diff with everything from before my change. We'd then discard that branch.
> 
> I can probably do all this Sunday or early next week, once I'm back.

Thanks. That would be great. Sorry for the huge churn here. There's no hurry on getting this merged. But I think it would be good to hold off any major harvesting until it's in (for instance, https://github.com/OneZoom/tree-build/pull/82/commits/e0963b234a3cd55585e63c1f128b3bb698a3b89a changes the logic which would otherwise overwrite old data).

File: oz_tree_build/newick/extract_trees.py [hunk 4]
```diff
@@ -24,86 +21,85 @@
...
-                if tree_string and tree_string[-1] == "(" and newick_tree[start] == ",":
-                    start += 1
-                return newick_tree[start:end]
+            tree_str = ""
 
             # Extract the subtree for this node, skipping over excluded ranges
-            prev_range = (node_start_index, node_start_index)
-            for range in excluded_ranges:
+            prev_r = (node_start_idx, node_start_idx)
+            for r in excluded_ranges:
```
Comment (davidebbo) [inline oz_tree_build/newick/extract_trees.py:90]: Are these renames like `prev_range -> prev_r` and `range -> r` enforced by Ruff?

File: oz_tree_build/newick/extract_trees.py [hunk 5]
```diff
@@ -24,86 +21,85 @@
...
-                # Exclude the comma before the excluded taxon. e.g. (A,B,REMOVE_ME) --> (A,B)
-                excluded_range = (node_start_index - 1, node_end_index)
-            elif newick_tree[node_end_index] == ",":
-                # Exclude the comma after the excluded taxon. e.g. (REMOVE_ME,B,C) --> (B,C)
-                excluded_range = (node_start_index, node_end_index + 1)
+            if newick_tree[node_start_idx - 1] == ",":
+                # Exclude comma before the excluded taxon. e.g. (A,B,REMOVE_ME) --> (A,B)
+                excluded_range = (node_start_idx - 1, node_end_idx)
+            elif newick_tree[node_end_idx] == ",":
+                # Exclude  comma after the excluded taxon. e.g. (REMOVE_ME,B,C) --> (B,C)
```
Comment (davidebbo) [inline oz_tree_build/newick/extract_trees.py:71]: A little strange how Ruff removed the `the` here, leaving 2 spaces. And aside from the spaces, it feels more semantically/grammatically correct with `the` in there.

File: oz_tree_build/newick/extract_trees.py (hunk 4 above)
Comment (hyanwong) [reply oz_tree_build/newick/extract_trees.py:90]: No, that was me making some changes. Ruff wasn't allowing "range" as a variable name because it conflicts with the python builtin `range` function. So I took the opportunity to shorted the variable name anyway (same with index -> idx). It's often a little easier to read with slightly shorter names, although perhaps `r` is a little too terse.

File: oz_tree_build/newick/extract_trees.py (hunk 5 above)
Comment (davidebbo) [reply oz_tree_build/newick/extract_trees.py:71]: I don't care strongly at all either way. It was all more in the context of reducing the diff to review. But that's ok, I'll just go through it as is. It's not so bad :)

Comment (hyanwong) [reply oz_tree_build/newick/extract_trees.py:71]: Yes, I made quite a lot of changes (>100, I think) in order to get all the ruff checks to pass. In retrospect, I suppose I should have put all the automatic changes into one commit, and all the hand-edited ones in another.

**Comment by davidebbo** [comment]
I need a scan through, but it's a bit large to fully review carefully. Is it fair to say that the majority of the changes are mechanical changes made by Ruff? Or are there also logic changes that you made outside of that? If there are certain areas that I should zoom in, let me know.

There is relatively good test coverage of the whole tree build process, so hopefully any significant regression would have been caught.

**Comment by hyanwong** [comment]
The vast majority of changes in https://github.com/OneZoom/tree-build/pull/82/commits/03f65241b9f3668dfe05f0b9b2661cd28c61f051 are simply line shortening or using format strings (where I have used "{s}" where s is the sql placeholder ("?" for sqlite, "%s" for mysql). One other common change is to make the docstring a single multi-line string, then taking the first para for the argpaser `help` parameter.

One thing that ruff picked up was the inadvisability of using mutable structures as default variables (as these get maintained between function calls). So instead of 

```
def foo(bar=[]):
```

It's better to do

```
def foo(bar=None):
    if bar is None:
        bar=[]
```

You might have noticed that in a previous PR I tried hard to remove the use of global variables such as "db" and pass them all as parameter values.

The only change I made to the expected results, that I recall, is in a different commit in this PR: https://github.com/OneZoom/tree-build/pull/82/commits/e0963b234a3cd55585e63c1f128b3bb698a3b89a

I'm pretty happy just to merge this and work on increasing the test coverage in case there are edge-cases that we hadn't thought about.

**Comment by hyanwong** [comment]
(Hopefully any further changes will be much more incremental.

**Comment by hyanwong** [comment]
Oh, one other thing I see now, which would be worth checking. As part of removing global variables: I make a single Azure Image  Client instance, and pass that around in the `crop` parameter. Alternatively, `crop` can be a tuple of 4 numbers, specifying a manual crop position. This means we can e.g. do bespoke cropping without asking Azure. At some point, we can also pass `None` to get a default (centre) crop. Testing the logic of this would be sensible, I guess.

**Comment by davidebbo** [comment]
Interesting, I would have expected `def foo(bar=[])` to create a new array each time no default is passed, but I see that's not the case. Not intuitive, but glad Ruff points this out.

And yes, I see you've made all kind of very nice clean up and improvements!

I'll take a look at the crop logic now.

**Comment by davidebbo** [comment]
For the cropping logic, it's not ideal to have two completely different data types passed in the same variables. It would be cleaner to just pass a lambda around to do the cropping. In one case it'd call Azure, and in the other it'd do center cropping, or whatever else.

I can make this change, but I wouldn't block on it now. I think we should get this PR in and move forward from there.

**Comment by hyanwong** [comment]
> For the cropping logic, it's not ideal to have two completely different data types passed in the same variables. It would be cleaner to just pass a lambda around to do the cropping. In one case it'd call Azure, and in the other it'd do center cropping, or whatever else.
> 
> I can make this change, but I wouldn't block on it now. I think we should get this PR in and move forward from there.

I'm not too keen about passing a lambda. But we could make a "crop" *class*, which we could initialise either to do the azure thing or to take a custom crop, I guess.

**Comment by hyanwong** [comment]
> It's the part about having to sniff the type with `if hasattr(crop, "analyze_from_url")` that gets dirty. Imagine if we had an alternate cloud service to experiment with, which would have a different client with a different shape. In general, the contract is: "given an image, give me a crop rectangle for that image", and there can be arbitrary implementations.

Yes, I agree. Using a wrapper class here would be cleaner, I think.

> Anyway, we can just leave it as is in the interest of time.

Agreed. We can replace this with something less dirty later.

**Comment by davidebbo** [comment]
Ok, I merged this PR.

I ended up reimplementing the crop using classes in a separate PR: https://github.com/OneZoom/tree-build/pull/89. PTAL.

**Comment by hyanwong** [comment]
Perfect, thanks *so* much!
